File: config/config_test.go
package config_test

import (
	"os"
	"strings"
	"testing"

	"github.com/truvity/policy/config"
)

// The configuration of the worked example, in miniature: the envelope the
// contract defines plus two fields of its own.
type shortener struct {
	Listen struct {
		Address string `json:"address"`
	} `json:"listen"`
	Probes struct {
		Address string `json:"address"`
	} `json:"probes"`
	Log struct {
		Level string `json:"level"`
	} `json:"log"`
	BaseURL  string `json:"baseURL"`
	Database struct {
		URL            string `json:"url"`
		PasswordEnv    string `json:"passwordEnv"`
		MaxConnections int    `json:"maxConnections"`
	} `json:"database"`
}

func schema(t *testing.T) []byte {
	t.Helper()
	b, err := os.ReadFile("testdata/shortener.schema.json")
	if err != nil {
		t.Fatal(err)
	}
	return b
}

func TestAValidFileLoadsAndDecodes(t *testing.T) {
	var cfg shortener
	if err := config.Load("testdata/valid.yaml", schema(t), &cfg); err != nil {
		t.Fatalf("a valid configuration was refused: %v", err)
	}

	// The fragments are resolved from the embedded copies, so a `$ref` to a
	// shape this repository publishes works with no network at all. If that
	// stopped being true, this would fail rather than quietly validating
	// nothing.
	if cfg.Probes.Address != ":7070" {
		t.Errorf("probes.address = %q, want \":7070\"", cfg.Probes.Address)
	}
	if cfg.Database.MaxConnections != 20 {
		t.Errorf("database.maxConnections = %d, want 20", cfg.Database.MaxConnections)
	}
	if cfg.BaseURL != "https://example.com" {
		t.Errorf("baseURL = %q", cfg.BaseURL)
	}
}

// The failure the whole contract exists to prevent: a key that means nothing,
// set by a deployment, doing nothing, with no signal but behaviour.
func TestAnUnknownKeyIsRefused(t *testing.T) {
	var cfg shortener
	err := config.Load("testdata/unknown-key.yaml", schema(t), &cfg)
	if err == nil {
		t.Fatal("a configuration with an unknown key was accepted")
	}
	if !strings.Contains(err.Error(), "databse") {
		t.Errorf("the error does not name the offending key, so nobody can fix it:\n%v", err)
	}
}

func TestAMissingRequiredKeyIsRefusedAndNamed(t *testing.T) {
	var cfg shortener
	err := config.Load("testdata/missing-required.yaml", schema(t), &cfg)
	if err == nil {
		t.Fatal("a configuration missing a required key was accepted")
	}
	if !strings.Contains(err.Error(), "database") {
		t.Errorf("the error does not name what is missing:\n%v", err)
	}
}

// "invalid config" is not an error message. The person reading it is looking
// at a file and needs the path.
func TestAWrongTypeNamesThePath(t *testing.T) {
	var cfg shortener
	err := config.Load("testdata/wrong-type.yaml", schema(t), &cfg)
	if err == nil {
		t.Fatal("a configuration with a string where a number belongs was accepted")
	}
	if !strings.Contains(err.Error(), "database.maxConnections") {
		t.Errorf("the error does not name the path that failed:\n%v", err)
	}
}

// A schema that is strict only at the top level accepts a secret smuggled
// into a nested object. This is the case the fragments' strictness exists
// for, and the error must not echo what it found.
func TestASecretInTheFileIsRefusedAndNotEchoed(t *testing.T) {
	var cfg shortener
	err := config.Load("testdata/secret-in-file.yaml", schema(t), &cfg)
	if err == nil {
		t.Fatal("a password written into the configuration file was accepted")
	}
	if strings.Contains(err.Error(), "hunter2") {
		t.Fatalf("the error quoted the value it refused, which puts it in every log that records the refusal:\n%v", err)
	}
	if !strings.Contains(err.Error(), "password") {
		t.Errorf("the error does not name the offending key:\n%v", err)
	}
}

func TestValidationHappensBeforeDecoding(t *testing.T) {
	// An invalid document must leave the caller's value untouched: a service
	// that half-decodes and then fails has a configuration nobody chose.
	cfg := shortener{BaseURL: "untouched"}
	_ = config.Load("testdata/wrong-type.yaml", schema(t), &cfg)
	if cfg.BaseURL != "untouched" {
		t.Errorf("the value was written to before validation failed: %q", cfg.BaseURL)
	}
}

func TestAMissingFileSaysWhichFile(t *testing.T) {
	var cfg shortener
	err := config.Load("testdata/there-is-no-such-file.yaml", schema(t), &cfg)
	if err == nil {
		t.Fatal("a missing configuration file was accepted")
	}
	if !strings.Contains(err.Error(), "there-is-no-such-file.yaml") {
		t.Errorf("the error does not name the file:\n%v", err)
	}
}

func TestAnEmptyFileIsRefused(t *testing.T) {
	empty := t.TempDir() + "/empty.yaml"
	if err := os.WriteFile(empty, nil, 0o600); err != nil {
		t.Fatal(err)
	}
	var cfg shortener
	if err := config.Load(empty, schema(t), &cfg); err == nil {
		t.Fatal("an empty configuration file was accepted, which would start the service on defaults nobody chose")
	}
}

func TestSecret(t *testing.T) {
	t.Run("set", func(t *testing.T) {
		t.Setenv("EXAMPLE_PASSWORD", "value")
		got, err := config.Secret("EXAMPLE_PASSWORD")
		if err != nil {
			t.Fatal(err)
		}
		if got != "value" {
			t.Errorf("got %q", got)
		}
	})

	t.Run("unset names the variable", func(t *testing.T) {
		_, err := config.Secret("EXAMPLE_PASSWORD_THAT_IS_NOT_SET")
		if err == nil {
			t.Fatal("an unset secret was accepted, so the service would start without it")
		}
		if !strings.Contains(err.Error(), "EXAMPLE_PASSWORD_THAT_IS_NOT_SET") {
			t.Errorf("the error does not name the variable:\n%v", err)
		}
	})

	t.Run("empty is not set", func(t *testing.T) {
		t.Setenv("EXAMPLE_EMPTY", "")
		if _, err := config.Secret("EXAMPLE_EMPTY"); err == nil {
			t.Fatal("an empty secret was accepted; an empty password is a misconfiguration, not a password")
		}
	})

	t.Run("no name at all", func(t *testing.T) {
		if _, err := config.Secret(""); err == nil {
			t.Fatal("a secret with no variable named was accepted")
		}
	})
}

File: ts/src/config.test.ts
/**
 * The same fixtures the Go loader is tested against, in the directory the Go
 * tests read them from.
 *
 * That is the point of this file, not an economy: two loaders that claim to
 * implement one contract must refuse the same documents and say something a
 * person can act on when they do. A fixture that only one of them sees is a
 * contract that exists twice.
 */
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { describe, expect, it } from "vitest";

import { ConfigError, load, secret, validate } from "./config.js";

const fixtures = fileURLToPath(new URL("../../config/testdata/", import.meta.url));
const schema = JSON.parse(readFileSync(`${fixtures}shortener.schema.json`, "utf8")) as object;
const fixture = (name: string) => `${fixtures}${name}`;

interface Shortener {
  probes: { address: string };
  baseURL: string;
  database: { url: string; passwordEnv?: string; maxConnections?: number };
}

describe("load", () => {
  it("accepts a valid file and returns it", () => {
    const cfg = load<Shortener>(fixture("valid.yaml"), schema);
    expect(cfg.probes.address).toBe(":7070");
    expect(cfg.database.maxConnections).toBe(20);
    expect(cfg.baseURL).toBe("https://example.com");
  });

  // The failure the whole contract exists to prevent: a key that means
  // nothing, set by a deployment, doing nothing, with no signal but
  // behaviour.
  it("refuses an unknown key and names it", () => {
    expect(() => load(fixture("unknown-key.yaml"), schema)).toThrow(/databse/);
  });

  it("refuses a missing required key and names it", () => {
    expect(() => load(fixture("missing-required.yaml"), schema)).toThrow(/database/);
  });

  // "invalid config" is not an error message. The person reading it is
  // looking at a file and needs the path.
  it("names the path when a type is wrong", () => {
    expect(() => load(fixture("wrong-type.yaml"), schema)).toThrow(/database\.maxConnections/);
  });

  it("refuses a password written into the file, and does not echo it", () => {
    let thrown: unknown;
    try {
      load(fixture("secret-in-file.yaml"), schema);
    } catch (error) {
      thrown = error;
    }
    expect(thrown).toBeInstanceOf(ConfigError);
    const message = (thrown as Error).message;
    expect(message).toMatch(/password/);
    expect(message).not.toMatch(/hunter2/);
  });

  it("says which file is missing", () => {
    expect(() => load(fixture("there-is-no-such-file.yaml"), schema)).toThrow(/there-is-no-such-file\.yaml/);
  });

  it("refuses an empty file", () => {
    // An empty file would otherwise start a service on defaults nobody chose.
    expect(() => load(fixture("empty.yaml"), schema)).toThrow(/is empty/);
  });

  // The wording is part of the contract: the same misconfiguration must read
  // the same way whichever runtime refused it.
  it("reports an unknown key the way the Go loader does", () => {
    let message = "";
    try {
      load(fixture("unknown-key.yaml"), schema);
    } catch (error) {
      message = (error as Error).message;
    }
    expect(message).toContain("databse: not a key this service reads");
  });
});

describe("validate", () => {
  it("accepts a document that satisfies the schema", () => {
    expect(() =>
      validate(
        {
          listen: { address: ":8080" },
          probes: { address: ":7070" },
          baseURL: "https://example.com",
          database: { url: "postgres://shortener@db:5432/shortener" },
        },
        schema,
      ),
    ).not.toThrow();
  });

  it("reports every failure at once, not one per run", () => {
    let failures: readonly string[] = [];
    try {
      validate({ baseURL: 7, database: { url: "not-a-postgres-url" } }, schema);
    } catch (error) {
      failures = (error as ConfigError).failures;
    }
    expect(failures.length).toBeGreaterThan(1);
  });
});

describe("secret", () => {
  it("reads the variable a configuration names", () => {
    process.env["EXAMPLE_PASSWORD"] = "value";
    expect(secret("EXAMPLE_PASSWORD")).toBe("value");
    delete process.env["EXAMPLE_PASSWORD"];
  });

  it("names the variable when it is unset", () => {
    expect(() => secret("EXAMPLE_PASSWORD_THAT_IS_NOT_SET")).toThrow(/EXAMPLE_PASSWORD_THAT_IS_NOT_SET/);
  });

  it("treats an empty variable as unset", () => {
    // An empty password is a misconfiguration, not a password.
    process.env["EXAMPLE_EMPTY"] = "";
    expect(() => secret("EXAMPLE_EMPTY")).toThrow(/is empty/);
    delete process.env["EXAMPLE_EMPTY"];
  });

  it("refuses a secret with no variable named", () => {
    expect(() => secret("")).toThrow();
  });
});
